### PR TITLE
Extra work

### PR DESCRIPTION
**Pull request by ChiefCharpa**
Need to put statistics in same class as game. Unsure on the testing.

### CODE REVIEW

File: CMP1903_A1_2324/Die.cs [hunk 1]
```diff
@@ -42,7 +42,7 @@ public void Roll()
                 * Due to random changing every clock tick a pause timer is required, 
                 * so before calling random a timer of 200 miliseconds is called to allow a new random number tyo be generated
                 */
-                int millisecond = 200;
+                int millisecond = 500;
                 Thread.Sleep(millisecond);
```
Comment (calmac04) [inline CMP1903_A1_2324/Die.cs]: creating an int for a set delay is unnecessary you can just put 1 in the brackets of thread.sleep to have the same effect

File: CMP1903_A1_2324/Program.cs [hunk 2]
```diff
@@ -10,10 +10,30 @@ internal class Program
...
+            bool valid = false;
+            while (!valid)
+            {
+                Console.WriteLine("Do you want to run the test: Y or N");
+                try
+                {
+                    string InputtedValue = Console.ReadLine();
+                    if (InputtedValue == "Y" || InputtedValue == "X")
+                    {
+
```
Comment (calmac04) [inline CMP1903_A1_2324/Program.cs]: code wont progress past this point so i am unable to verify the rest of the code as working

**Review by calmac04 — COMMENTED** [review]
overall code is of good quality and no blatant errors seem to exist outside of ones mentioned. primary focus should be on getting the user input in program to function so the rest of the code can be checked

**Review by Mr-X1214 — COMMENTED** [review]
In the die class Thread.Sleep() doesn't need 500, it could have a smaller number and still work. RollNumber could either be put in camelCase or _camelCase as it is a variable.

In the game class you could comment on why the code calculates the mean, median and mode as it's not clear and as far as I'm aware, not apart of the brief. It has good exception handling. Some variables probably should be changed to camelCase as they're currently in PascalCase

In the program class the comments are good and helpful. InputtedValue could be changed to camelCase as it seems to be a variable. Good error handling.

In the testing class you've used Debug.Assert() correctly. TestingRolls could be changed to camelCase but that's if the naming conventions still apply to arrays.

In summary, the code is generally easy to understand. It follows the style guide and fits the brief plus includes the additional feature that was stated.

File: CMP1903_A1_2324/Die.cs [hunk 3]
```diff
@@ -19,13 +19,13 @@ internal class Die
             //Property
 
             // Creates an encapsulated private variable for the number rolled by the Roll method
-            private int _RollNumber = 0;
+            private int _rollNumber = 0;
```
Comment (Slash1864) [inline CMP1903_A1_2324/Die.cs:22]: Correct naming of private/protected variable (using C# formatting guide)

File: CMP1903_A1_2324/Die.cs [hunk 4]
```diff
@@ -19,13 +19,13 @@ internal class Die
...
 
             // Creates the setter and getter for the private variable so it is encapsulated
-            public int RollNumber
+            public int rollNumber
```
Comment (Slash1864) [inline CMP1903_A1_2324/Die.cs:25]: Correct naming of local variable

File: CMP1903_A1_2324/Die.cs [hunk 5]
```diff
@@ -19,13 +19,13 @@ internal class Die
...
             {
-                get { return _RollNumber; }
-                set { _RollNumber = value; }
+                get { return _rollNumber; }
+                set { _rollNumber = value; }
```
Comment (Slash1864) [inline CMP1903_A1_2324/Die.cs:28]: Example of OOP technique encapsulation

File: CMP1903_A1_2324/Die.cs [hunk 6]
```diff
@@ -42,11 +42,11 @@ public void Roll()
...
-                Thread.Sleep(500);
+                Thread.Sleep(10);
                 // Instantiates an instance of the class random
                 Random rnd = new Random();
                 // Creates a random number between 1 and 6 for the dice roll and assigns the value to the private variable _RollNumber
-                RollNumber = rnd.Next(1, 7);
+                rollNumber = rnd.Next(1, 7);
             }
 
     }
```
Comment (Slash1864) [inline CMP1903_A1_2324/Die.cs:52]: Code is spaced out into appropriate chunks making it easy and simple to read.

File: CMP1903_A1_2324/Program.cs [hunk 7]
```diff
@@ -16,16 +16,16 @@ static void Main(string[] args)
             while (!valid)
             {
                 // A string variable InputtedValue is instantiated as an empty string
-                string InputtedValue = "";
+                string inputtedValue = "";
                 // An output is produced in the console window asking if the user wants to run a test
                 Console.WriteLine("Do you want to run the test: Y or N");
                 // Try is used to ensure the user inputs a valid input and prevents any possible crashes
                 try
```
Comment (Slash1864) [inline CMP1903_A1_2324/Program.cs:23]: Error handling is done correctly however it may be more user friendly to alert the user that they have inputted an incorrect letter/symbol and then prompt them if they want to run the test.

File: CMP1903_A1_2324/Program.cs [hunk 8]
```diff
@@ -16,16 +16,16 @@ static void Main(string[] args)
...
                 {
                     // The users input is taken and is attempted to be stored as a string in InputtedValue
-                    InputtedValue = Console.ReadLine();
+                    inputtedValue = Console.ReadLine();
```
Comment (Slash1864) [inline CMP1903_A1_2324/Program.cs:26]: Make use of the ToUpper() method to ensure that as long as the user inputs the correct character e.g. y/n then it still works as it can help avoid confusion or frustration.

File: CMP1903_A1_2324/Die.cs [hunk 9]
```diff
@@ -42,11 +42,11 @@ public void Roll()
                 * Due to random changing every clock tick a pause timer is required, 
                 * so before calling random a timer of 200 miliseconds is called to allow a new random number tyo be generated
                 */
-                Thread.Sleep(500);
+                Thread.Sleep(10);
                 // Instantiates an instance of the class random
                 Random rnd = new Random();
                 // Creates a random number between 1 and 6 for the dice roll and assigns the value to the private variable _RollNumber
-                RollNumber = rnd.Next(1, 7);
+                rollNumber = rnd.Next(1, 7);
```
Comment (Slash1864) [inline CMP1903_A1_2324/Die.cs:49]: Requirement 1: Die class generates a random number between 1 and 6

File: CMP1903_A1_2324/Game.cs [hunk 10]
```diff
@@ -146,28 +146,28 @@ public void RollingContinuous(bool Val, bool TestRun = false)
...
-                Console.WriteLine($"Dice {storedRolls.Count()} rolled a {DieRoll}");
+                Console.WriteLine($"Dice {_storedRolls.Count()} rolled a {DieRoll}");
 
                 // instantiates a variable to store the total of the dice
                 int Total = 0;
                 // iterates through the list adding each roll stored to the total variable
-                foreach (int previousRoll in storedRolls)
+                foreach (int previousRoll in _storedRolls)
                 {
                     Total = Total + previousRoll;
```
Comment (Slash1864) [inline CMP1903_A1_2324/Game.cs:172]: Requirement 2: Game class is created which successfully rolls 3 dice and randomly generates a number for each as well as adding them together at the end and displaying the value of the 3 rolls as well as the total value of the 3 added together.

File: CMP1903_A1_2324/Game.cs [hunk 11]
```diff
@@ -197,7 +197,9 @@ public void RollingContinuous(bool Val, bool TestRun = false)
                         // If the inputed value is an x Val is set to false ending the while loop 
                         if (InputtedValue == "x" || InputtedValue == "X")
                         {
-                            Val = false;
+                            val = false;
```
Comment (Slash1864) [inline CMP1903_A1_2324/Game.cs:200]: Requirement 2 Extra: Die rolls being performed continuously is successfully performed.

File: CMP1903_A1_2324/Testing.cs [hunk 12]
```diff
@@ -43,18 +46,26 @@ public void TestCode()
...
-            int[] TestingRolls = { 1, 2, 2, 2, 3, 4, 6, 6 };
+            int[] testingRolls = { 1, 2, 2, 2, 3, 4, 6, 6 };
 
             // Creates an instance of Statistics class
             // calls the Mean method passing a bool to run the testing code
-            game.Mean(TestingRolls,true);
+            game.Mean(testingRolls,true);
             // Calls and prints the Mode method along with passing booln variable to call the testing code
-            Console.WriteLine($"The current mode of the dice is: {game.Mode(TestingRolls, true)}");
+            Console.WriteLine($"The current mode of the dice is: {game.Mode(testingRolls, true)}");
```
Comment (Slash1864) [inline CMP1903_A1_2324/Testing.cs:61]: Requirement 3: Testing using debug.assert() to verify the numbers and the sum of them is successfully implemented and is easy for others to read and understand.

**Review by Slash1864 — COMMENTED** [review]
Final Points:
- Code is split up into micro chunks when associated with one another making it easy to read and understand. Also code is clearly commented on as to what each part work together to achieve.
- Code is formatted using the Styles Guideline correctly with the use of _rollNumber as an example for private properties and rollNumber as an example for local variables.
- All requirements + extras have been met with the code successfully using the die, game and testing classes resulting in an output for all 3 dice values, the total sum, mode and mean. Testing is implemented using debug.assert() successfully and die rolls can be done continuously.
- Error handling is also implemented successfully with users being asked to retype their input if their original input resulted in an error being produced or not being any of the intended input options.

**Comment by ClaudioIldefonso** [comment]
all requirements have been met and surpassed by Harrys code. the rolls works properly as well as giving a mean for the total roll. the style guidelines have been followed perfectly ad could not find a mistake. due to the large amount of comments there was no confusion and all code was easily understandable. due to the use of debug.assert all logic errors for the code are handled properly. great piece of code
